gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/LeaseAttemptStatus.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.orchestration;
+
+import java.io.IOException;
+
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.Getter;
+
+
+/*
+ Class used to encapsulate status of lease acquisition attempt and derivations should contain information specific to
+ the status that results. The #getDagAction and #getMinimumLingerDurationMillis are meant to be overriden and used by
+ relevant derived classes.
+ */
+public abstract class LeaseAttemptStatus {
+  public DagActionStore.DagAction getDagAction() {
+    return null;
+  }
+
+  public long getMinimumLingerDurationMillis() {
+    return 0;
+  }
+
+  public static class NoLongerLeasingStatus extends LeaseAttemptStatus {}
+
+  /*
+  The participant calling this method acquired the lease for the event in question. `Dag action`'s flow execution id
+  is the timestamp associated with the lease and the time the caller obtained the lease is stored within the
+  `leaseAcquisitionTimestamp` field. The `multiActiveLeaseArbiter` reference is used to recordLeaseSuccess for the
+  current LeaseObtainedStatus via the completeLease method from a caller without access to the {@link MultiActiveLeaseArbiter}.
+  */
+  @Data
+  public static class LeaseObtainedStatus extends LeaseAttemptStatus {
```

Comment:
it's atypical to have a derived class defined inside the class it's extending.  is this borne of a desire to keep these all together, rather than having a separate top-level file for each?

(I do agree that separating into four separate files would be far less comprehensible.)

honestly, what I liked about having these defined inside `MultiActiveLeaseArbiter` was that they lived all together (as they do here) AND that they together resided along side the abstraction that uses them (the MALA itself).

Reply:
I agree that it's much easier to understand these classes when lie together in one file. I am okay with them being separated from the MALA file since removing the prefix from every place they are used makes code more readable and concise.

Reply:
After thinking about the structures, here's my opinion,
It maybe (?) atypical for a derived class be inside the parent class, but it looks more atypical for an **interface** to contain an abstract parent class and base class.
Being able to write `LeaseAttemptStatus.LeaseObtainedStatus` looked like a determining factor for choosing this structure. This is like how enums are addressed too.